File: auth_service/routes/roles.py
"""
Roles Management Routes
=======================
CRUD operations for roles (IAM)
"""

import logging
import json
from fastapi import APIRouter, HTTPException, Header
from pydantic import BaseModel
from typing import Optional, List, Dict
from decimal import Decimal

from config.database import get_db_pool

router = APIRouter()
logger = logging.getLogger(__name__)


class RoleCreate(BaseModel):
    name: str
    description: Optional[str] = None
    mcp_access: List[str] = []
    tool_restrictions: Dict[str, Dict[str, List[str]]] = {}
    dashboard_access: str = "none"
    rate_limit: int = 100
    cost_limit_daily: Decimal = Decimal("10.00")
    token_expiry: int = 3600


class RoleUpdate(BaseModel):
    description: Optional[str] = None
    mcp_access: Optional[List[str]] = None
    tool_restrictions: Optional[Dict[str, Dict[str, List[str]]]] = None
    dashboard_access: Optional[str] = None
    rate_limit: Optional[int] = None
    cost_limit_daily: Optional[Decimal] = None
    token_expiry: Optional[int] = None


@router.get("/roles")
async def list_roles(x_user_role: str = Header(None, alias="X-User-Role")):
    """List all roles"""
    pool = await get_db_pool()
    
    try:
        async with pool.acquire() as conn:
            rows = await conn.fetch("""
                SELECT id, name, description, mcp_access, 
                       tool_restrictions, dashboard_access, rate_limit, 
                       cost_limit_daily, token_expiry, created_at, updated_at
                FROM auth_service.roles
                ORDER BY name
            """)
            
            # Parse tool_restrictions if it's a string
            roles = []
            for row in rows:
                role_dict = dict(row)
                if isinstance(role_dict.get('tool_restrictions'), str):
                    try:
                        role_dict['tool_restrictions'] = json.loads(role_dict['tool_restrictions'])
                    except:
                        role_dict['tool_restrictions'] = {}
                roles.append(role_dict)
            
            return {"roles": roles}
    
    except Exception as e:
        logger.error(f"Error listing roles: {e}")
        raise HTTPException(500, f"Failed to list roles: {str(e)}")


@router.get("/roles/{role_id}")
async def get_role(role_id: int):
    """Get role by ID"""
    pool = await get_db_pool()
    
    try:
        async with pool.acquire() as conn:
            row = await conn.fetchrow("""
                SELECT id, name, description, mcp_access, 
                       tool_restrictions, dashboard_access, rate_limit, 
                       cost_limit_daily, token_expiry, created_at, updated_at
                FROM auth_service.roles
                WHERE id = $1
            """, role_id)
            
            if not row:
                raise HTTPException(404, "Role not found")
            
            role_dict = dict(row)
            # Parse tool_restrictions if it's a string
            if isinstance(role_dict.get('tool_restrictions'), str):
                try:
                    role_dict['tool_restrictions'] = json.loads(role_dict['tool_restrictions'])
                except:
                    role_dict['tool_restrictions'] = {}
            
            return role_dict
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting role: {e}")
        raise HTTPException(500, f"Failed to get role: {str(e)}")


@router.post("/roles")
async def create_role(
    role: RoleCreate,
    x_user_role: str = Header(None, alias="X-User-Role")
):
    """Create new role (admin only)"""
    if x_user_role != "super_admin":
        raise HTTPException(403, "Admin access required")
    
    pool = await get_db_pool()
    
    try:
        async with pool.acquire() as conn:
            # Check if role exists
            existing = await conn.fetchval(
                "SELECT id FROM auth_service.roles WHERE name = $1",
                role.name
            )
            
            if existing:
                raise HTTPException(400, "Role already exists")
            
            # Create role
            role_id = await conn.fetchval("""
                INSERT INTO auth_service.roles 
                (name, description, mcp_access, tool_restrictions, 
                 dashboard_access, rate_limit, cost_limit_daily, token_expiry)
                VALUES ($1, $2, $3, $4::jsonb, $5, $6, $7, $8)
                RETURNING id
            """, role.name, role.description, role.mcp_access,
                json.dumps(role.tool_restrictions), role.dashboard_access, role.rate_limit,
                role.cost_limit_daily, role.token_expiry)
            
            logger.info(f"Created role: {role.name} (ID: {role_id})")
            return {"role_id": role_id, "name": role.name}
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating role: {e}")
        raise HTTPException(500, f"Failed to create role: {str(e)}")


@router.put("/roles/{role_id}")
async def update_role(
    role_id: int,
    role: RoleUpdate,
    x_user_role: str = Header(None, alias="X-User-Role")
):
    """Update role (admin only)"""
    if x_user_role != "super_admin":
        raise HTTPException(403, "Admin access required")
    
    pool = await get_db_pool()
    
    try:
        async with pool.acquire() as conn:
            # Build update query dynamically
            updates = []
            values = []
            param_count = 1
            
            if role.description is not None:
                updates.append(f"description = ${param_count}")
                values.append(role.description)
                param_count += 1
            
            if role.mcp_access is not None:
                updates.append(f"mcp_access = ${param_count}")
                values.append(role.mcp_access)
                param_count += 1
            
            if role.tool_restrictions is not None:
                updates.append(f"tool_restrictions = ${param_count}::jsonb")
                values.append(json.dumps(role.tool_restrictions))
                param_count += 1
            
            if role.dashboard_access is not None:
                updates.append(f"dashboard_access = ${param_count}")
                values.append(role.dashboard_access)
                param_count += 1
            
            if role.rate_limit is not None:
                updates.append(f"rate_limit = ${param_count}")
                values.append(role.rate_limit)
                param_count += 1
            
            if role.cost_limit_daily is not None:
                updates.append(f"cost_limit_daily = ${param_count}")
                values.append(role.cost_limit_daily)
                param_count += 1
            
            if role.token_expiry is not None:
                updates.append(f"token_expiry = ${param_count}")
                values.append(role.token_expiry)
                param_count += 1
            
            if not updates:
                raise HTTPException(400, "No fields to update")
            
            updates.append(f"updated_at = CURRENT_TIMESTAMP")
            values.append(role_id)
            
            query = f"""
                UPDATE auth_service.roles
                SET {', '.join(updates)}
                WHERE id = ${param_count}
            """
            
            result = await conn.execute(query, *values)
            
            if result == "UPDATE 0":
                raise HTTPException(404, "Role not found")
            
            logger.info(f"Updated role ID: {role_id}")
            return {"message": "Role updated"}
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating role: {e}")
        raise HTTPException(500, f"Failed to update role: {str(e)}")


@router.delete("/roles/{role_id}")
async def delete_role(
    role_id: int,
    x_user_role: str = Header(None, alias="X-User-Role")
):
    """Delete role (admin only)"""
    if x_user_role != "super_admin":
        raise HTTPException(403, "Admin access required")
    
    pool = await get_db_pool()
    
    try:
        async with pool.acquire() as conn:
            # Check if role is in use
            users_count = await conn.fetchval(
                "SELECT COUNT(*) FROM auth_service.users WHERE role_id = $1",
                role_id
            )
            
            if users_count > 0:
                raise HTTPException(400, f"Cannot delete role: {users_count} users assigned")
            
            result = await conn.execute(
                "DELETE FROM auth_service.roles WHERE id = $1",
                role_id
            )
            
            if result == "DELETE 0":
                raise HTTPException(404, "Role not found")
            
            logger.info(f"Deleted role ID: {role_id}")
            return {"message": "Role deleted"}
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting role: {e}")
        raise HTTPException(500, f"Failed to delete role: {str(e)}")

File: go/internal/admin/dal/roles.go
package dal

import (
	"context"
	"time"
)

// ── Row types ──────────────────────────────────────────────────────────────

// RoleRow is a tenant_roles row.
type RoleRow struct {
	ID          string    `json:"id"`
	TenantID    string    `json:"tenant_id"`
	Name        string    `json:"name"`
	DisplayName string    `json:"display_name"`
	Description string    `json:"description"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// GrantRow is a tenant_role_grants row.
type GrantRow struct {
	ID            string    `json:"id"`
	RoleID        string    `json:"role_id"`
	ApplicationID string    `json:"application_id"`
	CreatedAt     time.Time `json:"created_at"`
}

// MappingRow is a tenant_role_mappings row.
type MappingRow struct {
	ID        string    `json:"id"`
	TenantID  string    `json:"tenant_id"`
	Source    string    `json:"source"`
	Field     string    `json:"field"`
	Value     string    `json:"value"`
	RoleID    string    `json:"role_id"`
	CreatedAt time.Time `json:"created_at"`
}

// ── Role CRUD ──────────────────────────────────────────────────────────────

func (d *DB) ListRoles(ctx context.Context, tenantID string) ([]RoleRow, error) {
	rows, err := d.q.Query(ctx, `
		SELECT id::text, tenant_id::text, name, display_name, description, created_at, updated_at
		FROM them.tenant_roles
		WHERE tenant_id = $1
		ORDER BY name
	`, tenantID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []RoleRow
	for rows.Next() {
		var r RoleRow
		if err := rows.Scan(&r.ID, &r.TenantID, &r.Name, &r.DisplayName, &r.Description, &r.CreatedAt, &r.UpdatedAt); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	rows.Close()
	if out == nil {
		out = []RoleRow{}
	}
	return out, nil
}

func (d *DB) CreateRole(ctx context.Context, tenantID, name, displayName, description string) (RoleRow, error) {
	var r RoleRow
	err := d.q.QueryRow(ctx, `
		INSERT INTO them.tenant_roles (tenant_id, name, display_name, description)
		VALUES ($1, $2, $3, $4)
		RETURNING id::text, tenant_id::text, name, display_name, description, created_at, updated_at
	`, tenantID, name, displayName, description).
		Scan(&r.ID, &r.TenantID, &r.Name, &r.DisplayName, &r.Description, &r.CreatedAt, &r.UpdatedAt)
	return r, err
}

func (d *DB) GetRole(ctx context.Context, tenantID, roleID string) (RoleRow, error) {
	var r RoleRow
	err := d.q.QueryRow(ctx, `
		SELECT id::text, tenant_id::text, name, display_name, description, created_at, updated_at
		FROM them.tenant_roles
		WHERE tenant_id = $1 AND id = $2::uuid
	`, tenantID, roleID).
		Scan(&r.ID, &r.TenantID, &r.Name, &r.DisplayName, &r.Description, &r.CreatedAt, &r.UpdatedAt)
	return r, err
}

func (d *DB) UpdateRole(ctx context.Context, tenantID, roleID, name, displayName, description string) (RoleRow, error) {
	var r RoleRow
	err := d.q.QueryRow(ctx, `
		UPDATE them.tenant_roles
		SET name=$3, display_name=$4, description=$5, updated_at=now()
		WHERE tenant_id=$1 AND id=$2::uuid
		RETURNING id::text, tenant_id::text, name, display_name, description, created_at, updated_at
	`, tenantID, roleID, name, displayName, description).
		Scan(&r.ID, &r.TenantID, &r.Name, &r.DisplayName, &r.Description, &r.CreatedAt, &r.UpdatedAt)
	return r, err
}

func (d *DB) DeleteRole(ctx context.Context, tenantID, roleID string) error {
	return d.q.Exec(ctx, `
		DELETE FROM them.tenant_roles
		WHERE tenant_id = $1 AND id = $2::uuid
	`, tenantID, roleID)
}

// ── Grants ─────────────────────────────────────────────────────────────────

func (d *DB) ListGrants(ctx context.Context, tenantID, roleID string) ([]GrantRow, error) {
	rows, err := d.q.Query(ctx, `
		SELECT g.id::text, g.role_id::text, g.application_id::text, g.created_at
		FROM them.tenant_role_grants g
		JOIN them.tenant_roles r ON r.id = g.role_id
		WHERE r.tenant_id = $1 AND g.role_id = $2::uuid
		ORDER BY g.created_at
	`, tenantID, roleID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []GrantRow
	for rows.Next() {
		var g GrantRow
		if err := rows.Scan(&g.ID, &g.RoleID, &g.ApplicationID, &g.CreatedAt); err != nil {
			return nil, err
		}
		out = append(out, g)
	}
	rows.Close()
	if out == nil {
		out = []GrantRow{}
	}
	return out, nil
}

func (d *DB) AddGrant(ctx context.Context, tenantID, roleID, applicationID string) (GrantRow, error) {
	var g GrantRow
	err := d.q.QueryRow(ctx, `
		INSERT INTO them.tenant_role_grants (role_id, application_id)
		SELECT $2::uuid, $3::uuid
		FROM them.tenant_roles
		WHERE id = $2::uuid AND tenant_id = $1
		RETURNING id::text, role_id::text, application_id::text, created_at
	`, tenantID, roleID, applicationID).
		Scan(&g.ID, &g.RoleID, &g.ApplicationID, &g.CreatedAt)
	return g, err
}

func (d *DB) DeleteGrant(ctx context.Context, tenantID, roleID, grantID string) error {
	return d.q.Exec(ctx, `
		DELETE FROM them.tenant_role_grants g
		USING them.tenant_roles r
		WHERE g.id = $3::uuid AND g.role_id = $2::uuid AND r.id = g.role_id AND r.tenant_id = $1
	`, tenantID, roleID, grantID)
}

// ── Mappings ───────────────────────────────────────────────────────────────

func (d *DB) ListRoleMappings(ctx context.Context, tenantID, roleID string) ([]MappingRow, error) {
	rows, err := d.q.Query(ctx, `
		SELECT id::text, tenant_id::text, source, field, value, role_id::text, created_at
		FROM them.tenant_role_mappings
		WHERE tenant_id = $1 AND role_id = $2::uuid
		ORDER BY created_at
	`, tenantID, roleID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []MappingRow
	for rows.Next() {
		var m MappingRow
		if err := rows.Scan(&m.ID, &m.TenantID, &m.Source, &m.Field, &m.Value, &m.RoleID, &m.CreatedAt); err != nil {
			return nil, err
		}
		out = append(out, m)
	}
	rows.Close()
	if out == nil {
		out = []MappingRow{}
	}
	return out, nil
}

func (d *DB) AddRoleMapping(ctx context.Context, tenantID, roleID, source, field, value string) (MappingRow, error) {
	var m MappingRow
	err := d.q.QueryRow(ctx, `
		INSERT INTO them.tenant_role_mappings (tenant_id, source, field, value, role_id)
		SELECT $1, $3, $4, $5, $2::uuid
		FROM them.tenant_roles
		WHERE id = $2::uuid AND tenant_id = $1
		RETURNING id::text, tenant_id::text, source, field, value, role_id::text, created_at
	`, tenantID, roleID, source, field, value).
		Scan(&m.ID, &m.TenantID, &m.Source, &m.Field, &m.Value, &m.RoleID, &m.CreatedAt)
	return m, err
}

func (d *DB) DeleteRoleMapping(ctx context.Context, tenantID, roleID, mappingID string) error {
	return d.q.Exec(ctx, `
		DELETE FROM them.tenant_role_mappings
		WHERE id = $3::uuid AND role_id = $2::uuid AND tenant_id = $1
	`, tenantID, roleID, mappingID)
}
